### PR TITLE
drivers: hwinfo: Adds Ambiq hwinfo support

### PR DESCRIPTION
**Pull request by RichardSWheatley**
This commit adds support for the hwinfo for Apollo4P SoCs

Signed-off-by: Richard Wheatley <richard.wheatley@ambiq.com>

### CODE REVIEW

File: west.yml
Comment (aaronyegx) [inline west.yml:146]: Generally you should use HAL PR# instead of the commit# of your development branch before the changes merging the mainline. I reviewed the HAL changes and merged it. Please update the revision to the latest commit of hal_ambiq main branch.

File: drivers/hwinfo/hwinfo_ambiq.c
Comment (alexanderwachter) [inline drivers/hwinfo/hwinfo_ambiq.c:18]: Please put this struct inside the function and use the Zephyr naming conventions.

File: drivers/hwinfo/hwinfo_ambiq.c
Comment (alexanderwachter) [inline drivers/hwinfo/hwinfo_ambiq.c:35]: Is the byte-order correct?

File: drivers/hwinfo/hwinfo_ambiq.c
Comment (alexanderwachter) [inline drivers/hwinfo/hwinfo_ambiq.c:44]: Please use the Zephyr naming conventions.

File: drivers/hwinfo/hwinfo_ambiq.c
Comment (alexanderwachter) [inline drivers/hwinfo/hwinfo_ambiq.c:111]: How to clear the status then? Is it cleared automatically after a reboot/reset?

**Review by alexanderwachter — CHANGES_REQUESTED** [review]

**Comment by alexanderwachter** [comment]
@RichardSWheatley Thanks for the addition! Do you have a link to a datasheet describing the registers?

File: drivers/hwinfo/hwinfo_ambiq.c
Comment (RichardSWheatley) [reply drivers/hwinfo/hwinfo_ambiq.c:18]: Will do

File: drivers/hwinfo/hwinfo_ambiq.c
Comment (RichardSWheatley) [reply drivers/hwinfo/hwinfo_ambiq.c:35]: No, it is not. Will fix.

File: drivers/hwinfo/hwinfo_ambiq.c
Comment (RichardSWheatley) [reply drivers/hwinfo/hwinfo_ambiq.c:111]: A user cannot clear the status. That is why it returns -ENOSYS as other hwinfo drivers do.

**Comment by RichardSWheatley** [comment]
> @RichardSWheatley Thanks for the addition! Do you have a link to a datasheet describing the registers?

https://github.com/zephyrproject-rtos/hal_ambiq/blob/main/CMSIS/AmbiqMicro/Include/apollo4p.h

File: drivers/hwinfo/hwinfo_ambiq.c
Comment (RichardSWheatley) [reply drivers/hwinfo/hwinfo_ambiq.c:115]:
```
> 
>   union {
>     __IOM uint32_t STAT;      /*!< (@ 0x0000885C) This register contains the status for brownout
>                                                   events and the causes for resets.
>                                                   NOTE 1: All bits in this register, including
>                                                   reserved bits, are writable. Therefore care
>                                                   should be taken not to write this register.
>                                                   NOTE 2: This register is only reset by POI
>                                                   not by HRESETn. Its contents are intended
>                                                   to survive all reset level except POI and
>                                                   full power cycles.                                         */
```
> 
>     struct {
>       __IOM uint32_t EXRSTAT    : 1;            /*!< [0..0] Reset was initiated by an External Reset.                          */
>       __IOM uint32_t PORSTAT    : 1;            /*!< [1..1] Reset was initiated by a Power-On Reset.                           */
>       __IOM uint32_t BORSTAT    : 1;            /*!< [2..2] Reset was initiated by a Brown-Out Reset.                          */
>       __IOM uint32_t SWRSTAT    : 1;            /*!< [3..3] Reset was a initiated by SW POR or AIRCR Reset.                    */
>       __IOM uint32_t POIRSTAT   : 1;            /*!< [4..4] Reset was a initiated by Software POI Reset.                       */
>       __IOM uint32_t DBGRSTAT   : 1;            /*!< [5..5] Reset was a initiated by Debugger Reset.                           */
>       __IOM uint32_t WDRSTAT    : 1;            /*!< [6..6] Reset was initiated by a Watchdog Timer Reset.                     */
>       __IOM uint32_t BOUSTAT    : 1;            /*!< [7..7] An Unregulated Supply Brownout Event occured.                      */
>       __IOM uint32_t BOCSTAT    : 1;            /*!< [8..8] VDDC Analog Brownout Event occured.                                */
>       __IOM uint32_t BOFSTAT    : 1;            /*!< [9..9] VDDF Analog Brownout Event occured.                                */
>       __IOM uint32_t BOSSTAT    : 1;            /*!< [10..10] VDDS Analog Brownout Event occured.                              */
>             uint32_t            : 21;
>     } STAT_b;
>   } ;

File: drivers/hwinfo/hwinfo_ambiq.c
Comment (alexanderwachter) [inline drivers/hwinfo/hwinfo_ambiq.c:18]: Can you add a comment what trim_ver is?

Comment (RichardSWheatley) [reply drivers/hwinfo/hwinfo_ambiq.c:18]: Of course! I will add a code comment and make the variable name more appropriate.

To let you know, it is an internal trim revision for register visibility in user space and voltage trims from the factory.

I will try to capture that in better wording in the comments.

**Review by alexanderwachter — APPROVED** [review]

**Review by aaronyegx — APPROVED** [review]
Looks good to me.

**Comment by RichardSWheatley** [comment]
@carlescufi can this be merged? It has 2 approvers.

**Comment by alexanderwachter** [comment]
@RichardSWheatley it will be merged by somebody from the release team when they see it.
You have the approval from the maintainer (me) and a second approval.

**Pull request merged**
